### PR TITLE
riscv: move arch_is_user_contex() to a TLS-based implementation

### PR DESCRIPTION
**Pull request by npitre**
This is just goodness. My own crazy `arch_is_user_contex()` SMP code
is now gone and things are more efficient than ever even on non-SMP.

Except... maybe... for the QEMU bug this uncovered which took me way too
long to figure out.

### CODE REVIEW

**Comment by npitre** [comment]
After some delay due to yet another QEMU bug that needed figuring out, here the updated version rebased on latest main.

File: arch/riscv/core/fatal.c [hunk 1]
```diff
@@ -43,7 +43,6 @@ FUNC_NORETURN void z_riscv_fatal_error(unsigned int reason,
 #ifdef CONFIG_USERSPACE
 		LOG_ERR("     sp: " PR_REG, esf->sp);
 #endif
-		LOG_ERR("     tp: " PR_REG, esf->tp);
```
[the remark this answers is not in the document]
Comment (npitre) [reply arch/riscv/core/fatal.c:52]: the tp reg is set on task switch and is never (supposed to be) modified by regular code. So there is no point saving it.

File: arch/riscv/core/pmp.c [hunk 2]
```diff
@@ -207,7 +207,8 @@ static bool set_pmp_entry(unsigned int *index_p, uint8_t perm,
  */
 extern void z_riscv_write_pmp_entries(unsigned int start, unsigned int end,
 				      bool clear_trailing_entries,
-				      ulong_t *pmp_addr, ulong_t *pmp_cfg);
+				      const ulong_t *pmp_addr,
+				      const ulong_t *pmp_cfg);
```
Comment (edersondisouza) [inline arch/riscv/core/pmp.c:211]: I wonder how we maintain these workarounds, or better, how do we get rid of them? Once zephyr sdk gets a version of qemu with the fix?

**Review by edersondisouza — APPROVED** [review]
LGTM

File: arch/riscv/core/pmp.c (hunk 2 above)
Comment (npitre) [reply arch/riscv/core/pmp.c:211]: Either that or the Zephyr SDK incorporates the workaround in its QEMU version. Given the workaround is easy and functionality isn't impacted, I'd wait until the SDK gets a fixed QEMU and then revert those workarounds. In any case I'm keeping a virtual eye on it.

**Review by carlocaione — APPROVED** [review]

**Comment by carlescufi** [comment]
@kgugala @fkokosinski can you please review?

**Review by fkokosinski — APPROVED** [review]

**Comment by petermcs-01** [comment]
In the previous non-exception non-smp implementation, the flag was protected from being written to by the user task, is this still the case?

**Comment by npitre** [comment]
> In the previous non-exception non-smp implementation, the flag was protected
> from being written to by the user task, is this still the case?

No. However if the user task changes the flag to pretend being a kernel
task without having kernel privileges it will fault and be killed.

**Pull request merged**